backend/assignments/1/src/main/java/org/sanskar/ExecuteTransaction.java

```diff
@@ -0,0 +1,184 @@
+
+    public ExecuteTransaction(JsonNode transaction, CountDownLatch latch, List<Coin> coins, List<Trader> traders) {
+        this.transaction = transaction;
+        this.latch = latch;
+        this.coins = coins;
+        this.traders = traders;
+    }
+
+
+    @Override
+    public void run() {
+        try {
+            String transactionType = transaction.get("type").asText();
+            JsonNode jsondata = transaction.get("data");
+
+            switch (transactionType) {
+                case "BUY":
+                    executeBuyTransactions(jsondata);
+                    break;
+                case "SELL":
+                    executeSellTransactions(jsondata);
+                    break;
+                case "UPDATE_PRICE":
+                    executeUpdate_PriceTransactions(jsondata);
+                    break;
+                case "ADD_VOLUME":
+                    executeAdd_VolumeTransactions(jsondata);
+                    break;
+                default:
+                    break;
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        } finally {
+            // synchronisation utility allows for one thread to finish before others interrupt
+            latch.countDown();
+        }
+    }
+
+    public void executeBuyTransactions(JsonNode jsondata) {
+
+        // Retrieving data for a particular coin
+        String coindata = jsondata.get("coin").asText();
+        double reqQuantity = jsondata.get("quantity").asDouble();  // Use double for decimal values
+        String walletdata = jsondata.get("wallet_address").asText();
+
+        // Finding corresponding coin in the coin list
+        Coin targetCoin = coins.stream()
+                .filter(coin -> coin.getSymbol().equals(coindata))
+                .findFirst()
+                .orElse(null);
+
+        if (targetCoin != null) {
+
+            Lock lock = new ReentrantLock();
+            Condition pendingCondition = lock.newCondition();
+
+            // Thread safety using synchronized
+            synchronized (targetCoin) {
+
+                while (Double.parseDouble(targetCoin.getCirculatingSupply()) < reqQuantity) {
+                    try {
+                        LOGGER.info("Not enough volume for buy transactions. Waiting for volume to increase.");
+                        targetCoin.wait();
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    }
+                }
+                targetCoin.decreaseVolume((long) reqQuantity);
+                LOGGER.info("Transactions processed successfully for coin " + coindata);
+
+                // Updating trader's portfolio
+                Trader trader = traders.stream()
+                        .filter(trader1 -> trader1.getWalletAddress().equals(walletdata))
+                        .findFirst()
+                        .orElse(null);
+
+                if (trader != null) {
```

Comment:
What if trader is null? Since you already deducted the targetCoin volume and suppose trader walletAddress is invalid in that case coins are gone for forever